modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_lp_timer_zephyr.c

```diff
@@ -99,99 +61,65 @@ void nrf_802154_lp_timer_deinit(void)
 	nrf_802154_clock_lfclk_stop();
 }
 
-void nrf_802154_lp_timer_critical_section_enter(void)
-{
-	nrf_802154_sl_mcu_critical_state_t state;
-
-	nrf_802154_sl_mcu_critical_enter(state);
-
-	if (!m_in_critical_section) {
-		(void)z_nrf_rtc_timer_compare_int_lock(m_rtc_channel);
-		m_in_critical_section = true;
-	}
-
-	nrf_802154_sl_mcu_critical_exit(state);
-}
-
-void nrf_802154_lp_timer_critical_section_exit(void)
+uint64_t nrf_802154_platform_sl_lptimer_current_lpticks_get(void)
 {
-	nrf_802154_sl_mcu_critical_state_t state;
-
-	nrf_802154_sl_mcu_critical_enter(state);
-
-	m_in_critical_section = false;
-
-	z_nrf_rtc_timer_compare_int_unlock(m_rtc_channel, m_is_running);
-
-	nrf_802154_sl_mcu_critical_exit(state);
+	return z_nrf_rtc_timer_read();
 }
 
-uint32_t nrf_802154_lp_timer_time_get(void)
+uint64_t nrf_802154_platform_sl_lptimer_us_to_lpticks_convert(uint64_t us, bool round_up)
 {
-	/* Please note that this function does not handle RTC overflow */
-	return NRF_802154_SL_RTC_TICKS_TO_US(z_nrf_rtc_timer_read());
+	if (round_up) {
+		return NRF_802154_SL_US_TO_RTC_TICKS(us);
+	} else {
+		return NRF_802154_SL_US_TO_RTC_TICKS(us) - 1ULL;
+	}
 }
 
-uint32_t nrf_802154_lp_timer_granularity_get(void)
+uint64_t nrf_802154_platform_sl_lptimer_lpticks_to_us_convert(uint64_t lpticks)
 {
-	return NRF_802154_SL_US_PER_TICK;
+	/* Calculations are performed on uint64_t as it is safe to assume
+	 * overflow will not occur in any foreseeable future.
+	 */
+	return NRF_802154_SL_RTC_TICKS_TO_US(lpticks);
 }
 
-void nrf_802154_lp_timer_start(uint32_t t0, uint32_t dt)
+void nrf_802154_platform_sl_lptimer_schedule_at(uint64_t fire_lpticks)
 {
-	timer_start_at(m_rtc_channel, t0, dt);
+	/* This function is not required to be reentrant, hence no critical section. */
+	z_nrf_rtc_timer_set(m_rtc_channel, fire_lpticks, rtc_irq_handler, NULL);
 }
 
-bool nrf_802154_lp_timer_is_running(void)
+void nrf_802154_platform_sl_lptimer_disable(void)
 {
-	return m_is_running;
+	(void)z_nrf_rtc_timer_compare_int_lock(m_rtc_channel);
 }
 
-void nrf_802154_lp_timer_stop(void)
+void nrf_802154_platform_sl_lptimer_critical_section_enter(void)
 {
 	nrf_802154_sl_mcu_critical_state_t state;
 
 	nrf_802154_sl_mcu_critical_enter(state);
 
-	m_is_running = false;
-	(void)z_nrf_rtc_timer_compare_int_lock(m_rtc_channel);
+	m_critical_section_cnt++;
+
+	if (m_critical_section_cnt == 1UL) {
+		(void)z_nrf_rtc_timer_compare_int_lock(m_rtc_channel);
```

Comment:
Result of `z_nrf_rtc_timer_compare_int_lock` should be saved, say in `m_compare_int_lock_key`, and used together in
`nrf_802154_platform_sl_lptimer_critical_section_exit`

Reply:
Unfortunately not true. Consider lo priority code is here:
https://github.com/pkwiek/zephyr/blob/pr/extend-nrf-rtc-timer/drivers/timer/nrf_rtc_timer.c#L232
or in any place inside `z_nrf_rtc_timer_compare_int_lock` of **the nrf_rtc_timer.c** itself.

Consider this code is preempted by our hi prio interrupt, and our driver enters it's critical section (what includes `nrf_802154_platform_sl_lptimer_critical_section_enter`).
Then when our radio driver leaves its own critical section what includes `nrf_802154_platform_sl_lptimer_critical_section_exit`.
`z_nrf_rtc_timer_compare_int_unlock(key = true)` will be called blindly. Then return to interrupted https://github.com/pkwiek/zephyr/blob/pr/extend-nrf-rtc-timer/drivers/timer/nrf_rtc_timer.c#L232 and ooops !